pgx/src/shmem.rs

```diff
@@ -0,0 +1,159 @@
+use crate::pg_sys;
+use core::ops::{Deref, DerefMut};
+use std::cell::UnsafeCell;
+
+/// A Rust locking mechanism which uses a PostgreSQL LWLock to lock the data
+///
+/// This type of lock allows a number of readers or at most one writer at any
+/// point in time. The write portion of this lock typically allows modification
+/// of the underlying data (exclusive access) and the read portion of this lock
+/// typically allows for read-only access (shared access).
+///
+/// The lock is valid across processes as the LWLock is managed by Postgres. Data
+/// mutability once a lock is obtained is handled by Rust giving out & or &mut
+/// pointers.
+///
+/// When a lock is given out it is wrapped in a PgLwLockShareGuard or
+/// PgLwLockExclusiveGuard, which releases the lock on drop
+///
+/// # Poisoning
+/// This lock can not be poisoned from Rust. Panic and Abort are handled by
+/// PostgreSQL cleanly.
+
+pub struct PgLwLock<T> {
+    inner: UnsafeCell<Option<PgLwLockInner<T>>>,
+}
+
+impl<T> PgLwLock<T> {
+    /// Create a new lock for T by attaching a LWLock, which is looked up by name
+    pub fn new(lock: String, value: T) -> Self {
```

Comment:
Any particular reason why the lock name throughout is a `String` and not a `&str`?  

Maybe it should even be `&'static str`?